src/erlcloud_cognito.erl

```diff
@@ -0,0 +1,917 @@
+                             ProviderDetails, AttributeMapping, []).
+
+-spec update_identity_provider(binary(), binary(), map(), map(), list()) ->
+    {ok, map()} | {error, any()}.
+update_identity_provider(UserPoolId, ProviderName,
+                         ProviderDetails, AttributeMapping, IdpIdentifiers) ->
+    Config = erlcloud_aws:default_config(),
+    update_identity_provider(UserPoolId, ProviderName, ProviderDetails,
+                             AttributeMapping, IdpIdentifiers, Config).
+
+-spec update_identity_provider(binary(), binary(), map(), map(),
+                               list(), aws_config()) ->
+    {ok, map()} | {error, any()}.
+update_identity_provider(UserPoolId, ProviderName, ProviderDetails,
+                         AttributeMapping, IdpIdentifiers, Config) ->
+    Mandatory = #{
+        <<"UserPoolId">>       => UserPoolId,
+        <<"ProviderName">>     => ProviderName
+    },
+    Optional = #{
+        <<"ProviderDetails">>  => ProviderDetails,
+        <<"AttributeMapping">> => AttributeMapping,
+        <<"IdpIdentifiers">>   => IdpIdentifiers
+    },
+    request(Config, "UpdateIdentityProvider", make_request_body(Mandatory, Optional)).
+
+%%------------------------------------------------------------------------------
+%% Internal Functions
+%%------------------------------------------------------------------------------
+request(Config, Request) ->
+    Result = erlcloud_retry:request(Config, Request, fun handle_result/1),
+    case erlcloud_aws:request_to_return(Result) of
+        {ok, {_, <<>>}}     -> {ok, #{}};
+        {ok, {_, RespBody}} -> {ok, jsx:decode(RespBody, [return_maps])};
+        {error, _} = Error  -> Error
+    end.
+
+request(Config0, OperationName, Request) ->
+    case erlcloud_aws:update_config(Config0) of
+        {ok, Config} ->
+            Body       = jsx:encode(Request),
+            Operation  = "AWSCognitoIdentityProviderService." ++ OperationName,
+            Headers    = get_headers(Config, Operation, Body),
+            AwsRequest = #aws_request{service         = 'cognito-idp',
+                                      uri             = get_url(Config),
+                                      method          = post,
+                                      request_headers = Headers,
+                                      request_body    = Body},
+            request(Config, AwsRequest);
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+request_no_resp(Config, OperationName, Request) ->
+    case request(Config, OperationName, Request) of
+        {ok, _} -> ok;
+        Error   -> Error
+    end.
+
+make_request_body(Mandatory, Optional) ->
+    maps:merge(Mandatory, erlcloud_util:filter_empty_map(Optional)).
+
+get_headers(#aws_config{cognito_host = Host} = Config, Operation, Body) ->
+    Headers = [{"host", Host},
+               {"x-amz-target", Operation},
+               {"version", ?API_VERSION},
```

Comment:
According to this [doc](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/CommonParameters.html) both `Operation` and `version` should be query parameters, not headers.

Reply:
`host` and `x-amz-target` should come as headers

Reply:
this [pdf](https://docs.aws.amazon.com/cognitoidentity/latest/APIReference/cognito-identity-api.pdf) contains request examples with target as a header

not sure about version though.. do we really need the version? It wasn't used in other modules afaik, so probably just skip it. What's you thoughts?

Reply:
I have utilized this:
`erlcloud_aws:sign_v4(post, "/", Config, Headers, Body, Region, "cognito-idp", [{"version", ?API_VERSION}])` to send the version as QS and got this error back from aws:
**The request signature we calculated does not match the signature you provide**

If it is added to the header and use the:
`erlcloud_aws:sign_v4_headers(Config, Headers, Body, Region, "cognito-idp")` 
the request does return  the desired response.

Reply:
@arthurnovak it looks like `x-amx-target` header is used incorrectly. In this implementation it is always `AWSCognitoIdentityProviderService.DescribeIdentityProvider ` 
From your example:
> X-AMZ-TARGET: com.amazonaws.cognito.identity.model.AWSCognitoIdentityService.DescribeIdentityPool

It looks like we're using two separate service APIs here: [Amazon Cognito Federated Identities](https://docs.aws.amazon.com/cognitoidentity/latest/APIReference/API_DescribeIdentityPool.html) and [Amazon Cognito](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_AdminGetUser.html)

BTW, API version for Amazon Cognito Federated Identities is 2014-06-30

Reply:
If you referring to the create/describe/update identity providers like this:
[https://github.com/Amershan/erlcloud/blob/cognito/src/erlcloud_cognito.erl#L507](https://github.com/Amershan/erlcloud/blob/cognito/src/erlcloud_cognito.erl#L507)

These are the user pool identity providers not the federated ones. See [here](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_DescribeIdentityProvider.html)

Reply:
@kkuzmin it was just an example to show that `x-amz-target` can be used in headers and it definitely works not only for Cognito interface, but in other erlcloud modules. If we want to investigate if headers are used correctly further, i'm happy to be involved in that. However, it'll affect whole erlcloud and seems to be out of scope of this Cognito interface implementation.

as @Amershan mentioned above, this implementation covers only User Pools interface and not Identity Pools one (shouldn't be mixed up with identity providers, which belong to User Pools). Confusing aws naming :/

Reply:
@kkuzmin 
Renamed it to erlcloud_cognito_user_pools